content/courses/advanced-ada/chapters/access_types.rst

```diff
@@ -294,14 +294,402 @@ Unchecked Access
+    :class: ada-run-expect-failure
+
+    with Integer_Types; use Integer_Types;
+
+    procedure Show_Unchecked_Deallocation is
+
+       I, I_2 : Integer_Access;
+
+    begin
+       I := new Integer;
+
+       I_2 := I;
+
+       --  NOTE: I_2 points to the same
+       --        object as I.
+
+       Free (I);
+
+       --  NOTE: at this point, I_2 is a
+       --        dangling reference!
+
+       --  Calls to Free (I) are OK!
+
+       Free (I);
+       Free (I);
+
+       --  A call to Free (I_2) is
+       --  NOT OK:
+
+       Free (I_2);
+    end Show_Unchecked_Deallocation;
+
+As we've seen before, we can have multiple calls to :ada:`Free (I)`. However,
+the call to :ada:`I_2` fails because it points to an object that doesn't exist
+anymore.
+
+Because of these potential errors, it is the programmer's responsibility to be
+very careful when using unchecked deallocation and avoid creating dangling
+references.
+
+For the example we've just seen, a better approach could be to explicitly
+assign :ada:`null` to :ada:`I_2` to indicate that it doesn't point to any
+specific object:
+
+.. code:: ada run_button project=Courses.Advanced_Ada.Access_Types.Unchecked_Deallocation
+
+    with Integer_Types; use Integer_Types;
+
+    procedure Show_Unchecked_Deallocation is
+
+       I, I_2 : Integer_Access;
+
+    begin
+       I := new Integer;
+
+       I_2 := I;
+
+       --  NOTE: I_2 points to the same
+       --        object as I.
+
+       Free (I);
+
+       I_2 := null;
+
+       --  NOTE: I_2 doesn't point to any
+       --        object, so we can call Free.
+
+       Free (I_2);
+    end Show_Unchecked_Deallocation;
+
+Now, calling :ada:`Free (I_2)` doesn't cause any issues because it doesn't
+point to any object.
+
+
+Dereferencing dangling references
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Of course, we shouldn't try to dereference a dangling reference. For example:
+
+.. code:: ada run_button project=Courses.Advanced_Ada.Access_Types.Unchecked_Deallocation
+
+    with Ada.Text_IO;   use Ada.Text_IO;
+    with Integer_Types; use Integer_Types;
+
+    procedure Show_Unchecked_Deallocation is
+
+       I_1, I_2 : Integer_Access;
+
+    begin
+       I_1 := new Integer'(42);
+       I_2 := I_1;
+
+       Put_Line ("I_1 = " & Integer'Image (I_1.all));
+       Put_Line ("I_2 = " & Integer'Image (I_2.all));
+
+       Put_Line ("Freeing I_1");
+       Free (I_1);
+
+       if I_1 /= null then
+          Put_Line ("I_1 = " & Integer'Image (I_1.all));
+       end if;
+
+       if I_2 /= null then
+          Put_Line ("I_2 = " & Integer'Image (I_2.all));
+       end if;
+    end Show_Unchecked_Deallocation;
+
+In this example, we allocate an object for :ada:`I_1` and make :ada:`I_2` point
+to the same object. Note, however, that we call :ada:`Free` for :ada:`I_1`, but
+not for :ada:`I_2`. Therefore, after that call to :ada:`Free`,
+:ada:`I_1 = null`, but :ada:`I_2` still points to the object originally
+allocated for :ada:`I_1`.
```

Comment:
No, I_2 may or may not point to the originally allocated object after the call to Free to I_1. I_2 still has the non-null access value that formerly designated I_1.all, but I_1.all has been reclaimed by the call to Free and may be reused for other allocations. So I_2.all is non-null but what it points to is now undefined.
All we can say is that after the call to Free for I_1, I_2 remains non-null but becomes a dangling reference, and the effect of dereferencing I_2 is undefined. And really difficult to debug because the storage at the old address of I_1.all might not yet be used by another allocation, so it could look like I_1.all (and hence I2.all) is "still there" with the expected values.